docs/semantics/wrapped-errors.md

```diff
@@ -0,0 +1,117 @@
+
+For example, a warning attached to a value within a `Vector` can be wrapped in a
+`Map_Error` wrapper which indicates the position of the value within the
+`Vector`. 
+
+(Such errors are only wrapped when they are obtained through `Warning.get_all_wrapped`;
+see [Obtaining Wrapped Errors](#obtaining-wrapped-errors).)
+
+For example:
+
+```ruby
+err = My_Error.Error "my error"
+vec = [10, Warning.attach err 20, 30]
+IO.println (Warning.get_all vec)
+IO.println (Warning.get_all_wrapped vec)
+```
+
+Output:
+
+```ruby
+[My_Error.Error my error]
+[Map_Error.Error 1 (My_Error.Error my error)] # The error is at index 1
+```
+
+# Catching Wrapped Errors
+
+Wrapped errors are "transparent" to `Error.catch`. That is, if you attempt to
+catch a certain error, but the error that is actually thrown is wrapped, the
+catch will still succeed. In the example below, a `My_Error` error is thrown in
+a call to `map`, and so the resulting error is wrapped. You can catch the error
+as a `My_Error`, or as a `Map_Error`.
+
+```ruby
+fun x = if x == 20 then Error.throw (My_Error.Error "my error") else x
+[10, 20, 30].map fun . catch My_Error e->
+    IO.println e
+[10, 20, 30].map fun . catch Map_Error e->
+    IO.println e
+```
+
+Output:
+
+```ruby
+(My_Error.Error 'my error')
+(Map_Error.Error 1 (My_Error.Error 'my error'))
+```
+
+Note that if you catch the error as the inner error (`My_Error`), the
+`Map_Error` wrapper is stripped off.
+
+# Implementing Error Wrappers
+
+An error wrapper is a regular Enso value that has a conversion to `Wrapped_Error`. For example:
+
+```ruby
+type Map_Error
+    Error (index:Integer) inner_error
+
+Wrapped_Error.from (that : Map_Error) = Wrapped_Error.Value that that.inner_error
+```
+
+The `from` implementation allows `Error.catch` to detect that it is an error
+wrapper, and possibly perform automatic unwrapping on it.
+
+# Obtaining Wrapped Errors
+
+Wrapped errors are obtained in two ways:
+* An error thrown during a call to `Vector.map`
+* A warning attached to a value within a `Vector`
+
+In the case of an error thrown during `Vector.map`, the error is caught, wrapped
+in `Map_Error`, and re-thrown.
+
+In the case of a warning attached to a value within a `Vector`, the wrapper is
+added by `Warning.get_all_wrapped` when it is called on the `Vector`. In this
+case, the wrapping is not attached to the value itself, and is therefore not
+propagated to downstream values.
+
+# Map_Error
+
+`Map_Error` is the motivating example for wrapped errors, and is currently the
+only implemented error wrapper. It exists so that an error thrown during a `map`
+call over a large `Vector` will contain the index at which the error occurred,
+so it can be displayed to the user in the IDE.
+
+Note that the error does not have to occur during a call to `map`. A `Map_Error`
+wrapping is added by `Warning.get_all_wrapped` to any warning attached to a
+value within a `Vector` (or any array-like container). If the value is extracted
```

Comment:
> Note that the error does not have to occur during a call to `map`.

Does this suggest that the name `Map_Error` is actually misleading? Shall we choose some more general name that is more relevant to the more general use case of an error inside a vector?

(I'm not really sure what a good name for that would be... but `Map_Error` sounds very `map`-centrict, whereas we can see it is actually not)

Reply:
bump, as it seems this was not picked up?

I'm not advocating for definitely renaming it, but have we at least considered alternative names?